src/tools/crawler/crawler.rs

```diff
@@ -1,107 +1,100 @@
-            if let Payload::TmEndpoints(endpoints) = message.payload {
-                self.process_endpoints_message(addr, endpoints).await;
-            }
+            known_network: Arc::new(Default::default()),
         }
     }
+}
 
-    async fn process_endpoints_message(&mut self, addr: SocketAddr, endpoints: TmEndpoints) {
-        info!(
-            "Received endpoints from {}: {:?}",
-            addr, endpoints.endpoints_v2
-        );
-        for endpoint in endpoints.endpoints_v2 {
-            match SocketAddr::from_str(&endpoint.endpoint) {
-                Ok(peer) => {
-                    self.process_peer(addr, endpoint.hops, peer).await;
+/// Spawns a tokio's task to crawl given address. After receiving the response it will
+/// process it and start more crawl tasks recursively.
+pub(super) async fn crawl(addr: SocketAddr, known_network: Arc<KnownNetwork>) {
+    tokio::spawn(async move {
+        trace!("Crawling {}", addr);
+        if known_network.new_node(addr).await {
+            match get_crawl_response(addr).await {
+                Ok((response, duration)) => {
+                    process_response(addr, response, duration, known_network).await
                 }
                 Err(e) => {
-                    error!("invalid address from {}: {}", addr, e);
+                    warn!("Unable to get crawl response from {}: {:?}", addr, e);
+                    // TODO if it's connection refused or timeout: retry connection a few time after a while
                 }
             }
         }
-    }
-
-    async fn process_peer(&mut self, addr: SocketAddr, hops: u32, peer: SocketAddr) {
-        self.known_network.insert_node(peer).await;
-        if hops == 0 {
-            self.known_network.insert_connection(addr, peer).await;
-        }
-        self.get_peers(peer).await;
-    }
-
-    pub(super) async fn get_peers(&self, addr: SocketAddr) {
-        let node = self.node.clone();
-        let network = self.known_network.clone();
+    });
+}
 
-        tokio::spawn(async move {
-            if network.insert_node(addr).await {
-                let start = Instant::now();
-                if connect_node(node, addr).await {
-                    network.set_connected(addr, start.elapsed()).await;
-                }
+/// Processes /crawl response: updates `known_network` and starts more crawling recursively.
+fn process_response(
+    from: SocketAddr,
+    response: CrawlResponse,
+    connecting_time: Duration,
+    known_network: Arc<KnownNetwork>,
+) -> BoxFuture<'static, ()> {
+    // Wrapped in BoxFuture to allow for async recursion.
+    async move {
+        known_network
+            .update_stats(from, connecting_time, response.server.build_version)
+            .await;
+        for peer in response.overlay.active {
+            if let Some(addr) = parse_peer_addr(from, peer) {
+                known_network.insert_connection(from, addr).await;
```

Comment:
wouldn't it make sense to insert all the new connections based on the response first, and only then to crawl?

Reply:
I was thinking that we could insert connections wholesale, reducing lock contention, and only start crawling afterwards